README.md
OrangeTVKodi 2v1
============

Tato sada skriptu a doplňku pro Kodi slouží pro vygenerování playlistů tak, aby služba OrangeTV Slovensko šla použít v Kodi, nebo v TVHeadendu. Jedná sa o úpravu pôvodného projektu: [O2TVKodi](https://github.com/Pavuucek/O2TVKodi)  
Obě části jsou víceméně funkčně stejné a slouží k témuž účelu. Kterou část použijete záleží na vás.

Část 1 - Doplněk pro Kodi
-------------------------
Doplněk je vhodný zejména pro případy, kdy pro přehrávání používáte pouze IPTV Simple Client, nebo kdy je Kodi a TVHeadend spouštěn na jednom stroji.

### Instalace
- Stáhněte si [poslední verzi](https://github.com/cisarikj/OrangeTVKodi/releases/latest)
- Otevřete Kodi a povolte si **Neznámé zdroje** v **Nastavení -> Systém -> Doplňky** 
- Nainstalujte doplněk ze staženého zip souboru
- Proveďte nastavení doplňku - zadejte své uživatelské jméno a heslo. Doplněk najdete v sekci **Služby**

Pokud vše nastavíte správně, ve složce nastavení doplňku by se měly objevit soubory `orangetv.generic.m3u8` a `orangetv.playlist.m3u8`, které použijte pro IPTV Simple Client, nebo TVHeadend.
Umístění složky nastavení doplňku se liší podle platformy. Na windows bývá v C:\Users\jméno_uživatele\AppData\Roaming\Kodi\userdata\addon_data\service.playlist.orangetv\

Více informací naleznete na fóru [xbmc-kodi.cz](https://www.xbmc-kodi.cz/prispevek-playlist-orangetv-addon)

Část 2 - Samostatný skript
--------------------------
Skript je vhodný pro pokročilé uživatele, kteří mají aspoň základní znalosti Linuxu. Je vhodné ho spouštět periodicky (např. 1x za 24hod.)

### Instalace
- Nainstalujte si Python verze 2.7 (skript zatím není kompatibilní s Pythonem verze 3)
- Stáhněte si [poslední verzi](https://github.com/cisarikj/OrangeTVKodi/releases/latest) a rozbalte zip soubor do libovolné složky
- spusťte `pip install -r requirements.txt`, čímž si nainstalujete potřebné podpůrné knihovny
- upravte si nastavení v souboru `config.py`, zejména uživatelské jméno a heslo. Položku `device_id` můžete nechat prázdnou.
- spusťte skript pomocí příkazu `python playlist.py` nebo `./playlist.py`

Více informací naleznete na fóru [xbmc-kodi.cz](https://www.xbmc-kodi.cz/prispevek-playlist-orangetv-addon)

Řešení potíží
-------------
Nikdo není dokonalý a potíže se se mohou objevit kdykoliv.
V současné době je známo několik problémů na které byste mohli narazit:
- Nastavení skriptu verze 0.0.1 není kompatibilní s předchozími. Nachází se v jiném souboru, takže musíte zapsat veškerá nastavení znovu.
- V případě potíží při instalaci nové verze doplňku pro Kodi pomáhá smazat adresář s nastaveními. Na windows se nachází např. v `C:\Users\jméno_uživatele\AppData\Roaming\Kodi\userdata\addon_data\service.playlist.orangetv\`.
- Potíže s přihlášením zatím nejsou zcela vyřešeny a mohou záviset na problémech na straně Orange. 

Pokud jste objevili jiný problém, nebo chybu nejdříve hledejte v [seznamu chyb](https://github.com/Pavuucek/O2TVKodi/issues) a až v případě, že svojí chybu nenajdete, [vytvořte nové hlášení](https://github.com/cisarikj/OrangeTVKodi/issues/new).

Autoři
------

- Štěpán Ort
- JiRo
- Cromac
- Michal Kuncl (Pavůček)
- Jan Cisárik (johnyzh94) - úprava O2TVKodi pre OrangeTV

Pomoct s vývojem můžete i vy! Jak na to? Stačí si vytvořit [fork](https://help.github.com/articles/fork-a-repo) tohoto repozitáře, provést příslušné změny a poté otevřít [pull request](https://help.github.com/articles/using-pull-requests).


changelog.txt
2019-05-08 [0.1]
First release.



o2tvgo.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""Wrapper pro O2TV Go
"""

import requests

__author__ = "Štěpán Ort"
__license__ = "MIT"
__version__ = "1.1.8"
__email__ = "stepanort@gmail.com"

_COMMON_HEADERS = {"X-NanguTv-App-Version": "Android#1.2.9",
                   "X-NanguTv-Device-Name": "Nexus 7",
                   "User-Agent": "Dalvik/2.1.0 (Linux; U; Android 5.1.1; Nexus 7 Build/LMY47V)",
                   "Accept-Encoding": "gzip",
                   "Connection": "Keep-Alive"}


def _to_string(text):
    if type(text).__name__ == 'unicode':
        output = text.encode('utf-8')
    else:
        output = str(text)
    return output


# Kanál


class LiveChannel:

    # JiRo - doplněn parametr kvality
    def __init__(self, o2tv, channel_key, name, logo_url, weight, quality):
        self._o2tv = o2tv
        self.channel_key = channel_key
        self.name = name
        self.weight = weight
        self.logo_url = logo_url
        self.quality = quality  # doplněn parametr kvality

    def url(self):
        if not self._o2tv.access_token:
            self._o2tv.refresh_access_token()
        access_token = self._o2tv.access_token
        if not self._o2tv.subscription_code:
            self._o2tv.refresh_configuration()
        subscription_code = self._o2tv.subscription_code
        playlist = None
        while access_token:
            params = {"serviceType": "LIVE_TV",
                      "subscriptionCode": subscription_code,
                      "channelKey": self.channel_key,
                      "deviceType": self.quality,
                      "streamingProtocol": "HLS"}  # JiRo - doplněn parametr kvality
            headers = _COMMON_HEADERS
            cookies = {"access_token": access_token,
                       "deviceId": self._o2tv.device_id}
            req = requests.get('http://app01.gtm.orange.sk/sws/server/streaming/uris.json',
                               params=params, headers=headers, cookies=cookies)
            json_data = req.json()
            access_token = None
            if 'statusMessage' in json_data:
                status = json_data['statusMessage']
                if status == 'bad-credentials':
                    access_token = self._o2tv.refresh_access_token()
                elif status == 'channel.not-found':
                    raise ChannelIsNotBroadcastingError()
                else:
                    raise Exception(status)
            else:
                # Pavuucek: Pokus o vynucení HD kvality
                playlist = ""
                # pro kvalitu STB nebo PC se pokusíme vybrat HD adresu.
                # když není k dispozici, tak první v seznamu
                for uris in json_data["uris"]:
                    if self.quality == "STB" or self.quality == "PC":
                        if uris["resolution"] == "HD" and playlist == "":
                            playlist = uris["uri"]
                    else:
                        # pro ostatní vracíme SD adresu
                        if uris["resolution"] == "SD" and playlist == "":
                            playlist = uris["uri"]
                # playlist nebyl přiřazený, takže první adresa v seznamu
                if playlist == "":
                    playlist = json_data["uris"][0]["uri"]
        return playlist


class ChannelIsNotBroadcastingError(BaseException):
    pass


class AuthenticationError(BaseException):
    pass


class TooManyDevicesError(BaseException):
    pass


# JiRo - doplněna kontrola zaplacené služby
class NoPurchasedServiceError(BaseException):
    pass


class O2TVGO:

    def __init__(self, device_id, username, password, quality, log_function=None):  # JiRo - doplněn parametr kvality
        self.username = username
        self.password = password
        self._live_channels = {}
        self.access_token = None
        self.subscription_code = None
        self.locality = None
        self.offer = None
        self.device_id = device_id
        self.quality = quality  # JiRo - doplněn parametr kvality
        self.log_function = log_function

    def _log(self, message):
        if self.log_function:
            self.log_function(message)
    def get_access_token_password(self):
        self._log('Getting Token via password...')
        if not self.username or not self.password:
            raise AuthenticationError()
        headers = _COMMON_HEADERS
        headers["Content-Type"] = "application/x-www-form-urlencoded;charset=UTF-8"
        data = {'grant_type': 'password',
				'client_id': 'orangesk-mobile',
				'client_secret': 'e4ec1e957306e306c1fd2c706a69606b',
				'isp_id': '5',
                'username': self.username,
                'password': self.password,
                'platform_id': 'b0af5c7d6e17f24259a20cf60e069c22',
				'custom': 'orangesk-mobile',
				'response_type': 'token'
				}
        req = requests.post('https://oauth01.gtm.orange.sk/oauth/token',
                            data=data, headers=headers, verify=False)
        j = req.json()
        if 'error' in j:
            error = j['error']
            if error == 'authentication-failed':
                self._log('Authentication Error')
                return None
            else:
                raise Exception(error)
        self.access_token = j["access_token"]
        self.expires_in = j["expires_in"]
        self._log('Token OK')
        return self.access_token
		
    def refresh_access_token(self):
        if not self.access_token:
            self.get_access_token_password()
        if not self.access_token:
            self._log('Authentication Error (failed to get token)')
            raise AuthenticationError()
        return self.access_token

    def refresh_configuration(self):
        if not self.access_token:
            self.refresh_access_token()
        access_token = self.access_token
        headers = _COMMON_HEADERS
        cookies = {"access_token": access_token, "deviceId": self.device_id}
        req = requests.get(
            'https://app01.gtm.orange.sk/sws//subscription/settings/subscription-configuration.json', headers=headers,
            cookies=cookies)
        j = req.json()
        if 'errorMessage' in j:
            error_message = j['errorMessage']
            status_message = j['statusMessage']
            # JiRo - změna z 'unauthorized-device' na 'devices-limit-exceeded'
            if status_message == 'devices-limit-exceeded':
                raise TooManyDevicesError()
            else:
                raise Exception(error_message)
        self.subscription_code = _to_string(j["subscription"])
        self.offer = j["billingParams"]["offers"]
        self.tariff = j["billingParams"]["tariff"]
        self.locality = j["locality"]

    def live_channels(self):
        if not self.access_token:
            self.refresh_access_token()
        access_token = self.access_token
        if not self.offer:
            self.refresh_configuration()
        offer = self.offer
        if not self.tariff:
            self.refresh_configuration()
        tariff = self.tariff
        if not self.locality:
            self.refresh_configuration()
        locality = self.locality
        quality = self.quality  # JiRo - doplněn parametr kvality
        if len(self._live_channels) == 0:
            headers = _COMMON_HEADERS
            cookies = {"access_token": access_token,
                       "deviceId": self.device_id}
            params = {"locality": self.locality,
                      "tariff": self.tariff ,
                      "isp": "5",
					  "imageSize": "LARGE",
                      "language": "slo",
                      "deviceType": "PC",
                      "liveTvStreamingProtocol": "HLS",
                      "offer": self.offer}  # doplněn parametr kvality
            req = requests.get('http://app01.gtm.orange.sk/sws/server/tv/channels.json',
                               params=params, headers=headers, cookies=cookies)
            j = req.json()
            purchased_channels = j['purchasedChannels']
            if len(purchased_channels) == 0:  # JiRo - doplněna kontrola zaplacené služby
                raise NoPurchasedServiceError()  # JiRo - doplněna kontrola zaplacené služby
            items = j['channels']
            for channel_id, item in items.iteritems():
                if channel_id in purchased_channels:
                    live = item['liveTvPlayable']
                    if live:
                        channel_key = _to_string(item['channelKey'])
                        logo = _to_string(item['logo'])
                        if not logo.startswith('http://'):
                            logo = 'http://app01.gtm.orange.sk/' + logo
                        name = _to_string(item['channelName'])
                        weight = item['weight']
                        self._live_channels[channel_key] = LiveChannel(
                            self, channel_key, name, logo, weight, quality)  # doplněn parametr kvality
            done = False
            offset = 0     
                           
        return self._live_channels


playlist.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-


# Script pro generaci playlistu z OTA služby OrangeTV
# ***********************************************
# Script je odvozen z Kodi addon service.playlist.o2tv,
# který byl vytvořen z původního addon autora Štěpána Orta.


import os
import time

import urllib3

import common as c
import config as cfg
from o2tvgo import AuthenticationError
from o2tvgo import ChannelIsNotBroadcastingError
from o2tvgo import NoPurchasedServiceError
from o2tvgo import O2TVGO
from o2tvgo import TooManyDevicesError

urllib3.disable_warnings()


def _cut_log(limit, reduction):
    if cfg.cut_log == 0:
        return
    try:
        f = open(c.log_file, 'r')
        lines = f.readlines()
        f.close()
    except IOError:
        return
    else:
        length = len(lines)
        count = 0
        if length > limit:
            limit = length - limit + reduction + 1
            new_lines = ''
            for line in lines:
                count += 1
                if count < limit:
                    continue
                new_lines += line
            f = open(c.log_file, 'w')
            f.write(new_lines)
            f.close()
        return


def _log(message):
    f = open(c.log_file, 'a')
    message = format('%s %s' % (time.strftime('%Y-%m-%d %H:%M:%S'), message))
    print(message)
    f.write(message + "\n")
    f.close()


def _get_id(name):
    _id = ''
    try:
        f = open(name, 'r')
        lines = f.readlines()
    except IOError:
        return _id
    else:
        _id = lines[0].rstrip()
        f.close()
        return _id


def check_config():
    if cfg.username == '' or cfg.password == '':
        return False
    return True


def _fetch_channels():
    global _o2tvgo_
    channels = None
    while not channels:
        try:
            channels = _o2tvgo_.live_channels()
        except AuthenticationError:
            return None, c.authent_error
        except TooManyDevicesError:
            return None, c.toomany_error
        except NoPurchasedServiceError:
            return None, c.nopurch_error
    return channels, 'OK'


def _logo_file(channel):
    if cfg.channel_logo_name == 0:
        f = c.logo_name(channel) + '.png'
    elif cfg.channel_logo_name == 1:
        f = c.logo_name(channel) + '.jpg'
    elif cfg.channel_logo_name == 2:
        f = channel + '.png'
    elif cfg.channel_logo_name == 3:
        f = channel + '.jpg'
    else:
        return ''
    return f


def _logo_path_file(channel):
    if cfg.channel_logo == 2:
        path_file = os.path.join(cfg.channel_logo_path, _logo_file(channel))
        if not os.path.isfile(path_file):
            path_file = ""
    elif cfg.channel_logo == 3:
        path_file = cfg.channel_logo_url + _logo_file(channel)
    elif cfg.channel_logo == 4:
        path_file = c.marhy[cfg.channel_logo_github] + c.logo_name(channel) + '.png'
    else:
        return ''
    return path_file


def channel_playlist():
    channels, _code = _fetch_channels()
    if not channels:
        return _code, -1, -1
    channels_sorted = sorted(channels.values(), key=lambda _channel: _channel.weight)
    if cfg.channel_group == 1:
        group = c.default_group_name
    else:
        group = cfg.channel_group_name
    if cfg.my_script == 1:
        streamer = c.pipe + os.path.join(cfg.playlist_path, cfg.my_script_name)
    else:
        streamer = c.pipe + os.path.join(cfg.playlist_path, cfg.playlist_streamer)
    playlist_src = '#EXTM3U\n'
    playlist_dst = '#EXTM3U\n'
    _num = 0
    _err = 0
    for channel in channels_sorted:
        try:
            _log("Adding: " + c.to_string(channel.name))
            playlist_src += '#EXTINF:-1, %s\n%s\n' % (c.to_string(channel.name), c.to_string(channel.url()))
            playlist_dst += c.build_channel_lines(channel, cfg.channel_logo ,_logo_path_file(channel.name), streamer, group, cfg.playlist_type, cfg.channel_epg_name, cfg.channel_epg_id, cfg.channel_group)
            _num += 1
        except ChannelIsNotBroadcastingError:
            _err += 1
            _log("... Not broadcasting. Skipped.")
        except AuthenticationError:
            return c.authent_error, 0, 0
        except TooManyDevicesError:
            return c.toomany_error, 0, 0
    c.write_file(playlist_src, os.path.join(cfg.playlist_path, cfg.playlist_src), _log)
    c.write_file(playlist_dst, os.path.join(cfg.playlist_path, cfg.playlist_dst), _log)
    return 'OK', _num, _err


_cut_log(cfg.log_limit, cfg.log_reduction)
_log("--------------------")
_log('OrangeTVKodi Playlist')
_log('Version: %s %s' % (c.version, c.date))
_log("--------------------")
_log("Starting...")

if not check_config():
    _log('Invalid username or password.')
    _log('Please check config.py')
    exit()
_log('Config OK')
device_id = _get_id(c.id_file)
if not (device_id or cfg.device_id):
    first_device_id = c.device_id()
    second_device_id = c.device_id()
    if first_device_id == second_device_id:
        cfg.device_id = first_device_id
    else:
        _device_id_ = c.random_hex16()
    _log('New Device Id: %s' % cfg.device_id)
else:
    if device_id:
        cfg.device_id = device_id
c.write_file(cfg.device_id, c.id_file, _log)

if cfg.stream_quality == 'PC':
    _quality_ = 'PC'
else:
    _quality_ = 'MOBILE'

_o2tvgo_ = O2TVGO(cfg.device_id, cfg.username, cfg.password, _quality_, _log)

if cfg.playlist_type == 3:
    c.write_streamer(os.path.join(cfg.playlist_path, cfg.playlist_streamer),
                     os.path.join(cfg.playlist_path, cfg.playlist_src),
                     cfg.ffmpeg_command, _log)

code, num, err = channel_playlist()

_log('Download done with result EXIT: %s , DOWNLOADED: %d, SKIPPED: %d' % (code, num, err))
_log('Finished')


common.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import os
import random
import stat
import unicodedata
from uuid import getnode as get_mac

import config as cfg

version = '0.1'
date = '2019-05-08'
pipe = 'pipe://'
default_group_name = "OrangeTV"
marhy = 'https://marhycz.github.io/picons/640/', 'https://marhycz.github.io/picons/1024/'
log_file = os.path.join(cfg.playlist_path, 'playlist.log')
id_file = os.path.join(cfg.playlist_path, 'device_id')
authent_error = 'AuthenticationError'
toomany_error = 'TooManyDevicesError'
nopurch_error = 'NoPurchasedServiceError'


def device_id():
    mac = get_mac()
    hexed = hex((mac * 7919) % (2 ** 64))
    return ('0000000000000000' + hexed[2:-1])[16:]


def random_hex16():
    return ''.join([random.choice('0123456789abcdef') for x in range(16)])


def to_string(text):
    if type(text).__name__ == 'unicode':
        output = text.encode('utf-8')
    else:
        output = str(text)
    return output


def logo_name(channel):
    channel = unicode(channel, 'utf-8')
    channel = unicodedata.normalize('NFKD', channel)
    channel = channel.lower()
    name = ''
    for char in channel:
        if not unicodedata.combining(char) and (char.isalpha() or char.isdigit()):
            name += char
    return name


def add_param(param, value, cond):
    item = ''
    if cond:
        item = ' %s="%s"' % (param, str(value))
    return item


def write_file(content, name, log=None):
    if not log is None:
        log("Saving file: " + name)
    f = open(name, 'w')
    f.write(content)
    f.close()


def try_exec(name):
    f = name
    try:
        sts = os.stat(f)
        if not (sts.st_mode & stat.S_IEXEC):
            os.chmod(f, sts.st_mode | stat.S_IEXEC)
    except:
        pass


def write_streamer(streamer_file, playlist_file, ffmpeg_cmd, log=None):
    streamer_code = '#! /bin/bash\n' + \
                    'source=$*\n' + \
                    'tempplaylist=$(mktemp -u)".m3u8"\n' + \
                    'stream=$(grep -A 1 "${source}$" ' + playlist_file + \
                    ' | head -n 2 | tail -n 1)\n' + \
                    'wget -qO ${tempplaylist} ${stream}\n' + \
                    'streamcount=$(cat ${tempplaylist} | grep -Eo "(http|https)://[\da-z./?A-Z0-9\D=_-]*" | wc -l)\n' + \
                    'streamcount=$((streamcount-1))\n' + \
                    'if  [ "$streamcount" = "-1" ]; then streamcount=0; fi\n' + \
                    ffmpeg_cmd + ' -protocol_whitelist file,http,https,tcp,tls -fflags +genpts ' + \
                    '-loglevel fatal -i ${tempplaylist} -probesize 32 -reconnect_at_eof 1 -reconnect_streamed 1 ' + \
                    '-c copy -map p:${streamcount}? -f mpegts -tune zerolatency -bsf:v h264_mp4toannexb,dump_extra ' + \
                    '-mpegts_service_type digital_tv pipe:1\n'

    if not log is None:
        log('Saving Streamer: ' + streamer_file)
    write_file(streamer_code, streamer_file + '.sample')
    # _to_file(c.streamer_code, os.path.join(cfg.playlist_path, cfg.playlist_streamer + '.sample'))
    try_exec(streamer_file + '.sample')
    write_file(streamer_code, streamer_file)
    try_exec(streamer_file)


def build_channel_lines(channel, channel_logo, logoname, streamer, group, playlist_type, channel_epg_name, channel_epg_id, channel_group):
    name = channel.name
    logo = to_string(channel.logo_url)
    url = to_string(channel.url())
    epgname = name
    epgid = name
    r = ""
    # číslo programu v epg
    # viz https://www.o2.cz/file_conver/174210/_025_J411544_Razeni_televiznich_programu_O2_TV_03_2018.pdf
    channel_weight = to_string(channel.weight)
    # logo v mistnim souboru - kdyz soubor neexistuje, tak pouzit url
    if (channel_logo > 1) and (logoname != ""):
        logo = logoname
    if playlist_type == 1:
        r += '#EXTINF:-1'
        r += add_param('tvg-name', epgname, channel_epg_name != 0)
        r += add_param('tvg-id', epgid, channel_epg_id != 0)
        r += add_param('tvg-logo', logo, channel_logo != 0)
        r += add_param('tvg-chno', channel_weight, channel_epg_id != 0)
        r += add_param('group-titles', group, channel_group != 0)
        r += ', %s\n%s\n' % (name, url)
    if (playlist_type == 2) or (playlist_type == 3):
        r += '#EXTINF:-1'
        r += add_param('tvg-id', epgid, channel_epg_id != 0)
        r += add_param('tvg-logo', logo, channel_logo != 0)
        r += add_param('tvg-chno', channel_weight, channel_epg_id != 0)
        r += add_param('group-titles', group, channel_group != 0)
        r += ', %s\n' % name
        if playlist_type == 2:
            r += '%s\n' % url
        if playlist_type == 3:
            r += '%s %s\n' % (streamer, name)
    return r